be/src/storage/rowset/json_column_writer.cpp

```diff
@@ -0,0 +1,342 @@
+          _json_column_writer(std::move(json_writer)),
+          _json_meta(opts.meta),
+          _wfile(wfile) {}
+
+Status FlatJsonColumnWriter::append(const Column& column) {
+    return Status::InternalError("json column writer not support append reference interface");
+}
+
+Status FlatJsonColumnWriter::append(const ColumnPtr& column) {
+    _json_datas.emplace_back(column);
+    return Status::OK();
+}
+
+void FlatJsonColumnWriter::_flat_column(std::vector<ColumnPtr>& json_datas) {
+    DCHECK(_flat_paths.empty());
+    DCHECK(!json_datas.empty());
+
+    size_t total_rows = 0;
+    size_t null_count = 0;
+
+    for (auto& column : json_datas) {
+        DCHECK(!column->is_constant());
+
+        total_rows += column->size();
+        if (column->only_null() || column->empty()) {
+            null_count += column->size();
+            continue;
+        } else if (column->is_nullable()) {
+            auto* nullable_column = down_cast<NullableColumn*>(column.get());
+            null_count += nullable_column->null_count();
+        }
+    }
+
+    // more than half of null
+    if (null_count > total_rows * config::json_flat_null_factor) {
+        VLOG(8) << "flat json, null_count[" << null_count << "], row[" << total_rows
+                << "], null_factor: " << config::json_flat_null_factor;
+        return;
+    }
+
+    // extract common keys
+    std::unordered_map<std::string, uint64_t> hit_maps;
+    for (size_t k = 0; k < json_datas.size(); k++) {
+        size_t row_count = json_datas[k]->size();
+
+        ColumnViewer<TYPE_JSON> viewer(json_datas[k]);
+        for (size_t i = 0; i < row_count; ++i) {
+            if (viewer.is_null(i)) {
+                continue;
+            }
+
+            JsonValue* json = viewer.value(i);
+            auto vslice = json->to_vslice();
+
+            if (vslice.isNone() || vslice.isNull()) {
+                continue;
+            }
+
+            if (!vslice.isObject()) {
+                VLOG(8) << "flat json, row isn't object, can't be flatten";
+                return;
+            }
+
+            std::vector<std::string> keys = arangodb::velocypack::Collection::keys(json->to_vslice());
+            for (auto& sr : keys) {
+                hit_maps[sr]++;
+            }
+        }
+    }
+
+    if (hit_maps.size() <= config::json_flat_internal_column_min_limit) {
+        VLOG(8) << "flat json, internal column too less: " << hit_maps.size()
+                << ", at least: " << config::json_flat_internal_column_min_limit;
+        return;
+    }
+
+    // sort by hit
+    std::vector<pair<std::string, std::uint64_t>> top_hits(hit_maps.begin(), hit_maps.end());
+    std::sort(top_hits.begin(), top_hits.end(),
+              [](const pair<std::string, std::uint64_t>& a, const pair<std::string, std::uint64_t>& b) {
+                  return a.second > b.second;
+              });
+
+    for (int i = 0; i < top_hits.size() && i < config::json_flat_column_max; i++) {
```

Comment:
it's more reasonable for make it a table-granular configuration

Reply:
I know, but it's hard now, it's worked before flattern json, we need save whole table data in memory, the current design allows different schemas between segments.

And in future, we can't to make it as a table-granular configuration in FE, by analyzing the usage frequency of subfield in the query